internal/consumer/service.go

```diff
@@ -23,19 +23,23 @@ type Client interface {
 	processStories(ctx context.Context)
 }
 
-func NewService(logger *zap.Logger, config *model.Configuration, hnClient hackernews.Client, dbClient database.Client) (*service, error) {
-	if hnClient == nil {
+func NewService(logger *zap.Logger, config *model.Configuration, dbClient database.Client, hnClient ...hackernews.Client) (*service, error) {
```

Comment:
The Functional Options approach is slightly different than using variadic arguments. https://www.sohamkamani.com/golang/options-pattern/#defining-the-functional-options this article is a bit more succinct than the one I linked originally.

Essentially you define a type that's a function which accepts the type you want to modify e.g.

`type ServiceOption func (*service)`

and then accept a variadic argument of that type:

```
func NewService(logger *zap.Logger, config *model.Configuration, dbClient database.Client, ...opts ServiceOption) (*service, error) {
    s := &service{
        // add defaults here
    }

    // here we use the functional options to override defaults
    for _, opt := range opts {
        opt(s)
    }
}
```

You can then define specific functional options to configure different parts of the service's defaults

```
func WithHttpClient(client http.Client) ServiceOption {
    return func(s *service) {
        s.client = client
    }
}       
```

```
// another example just to demo
func WithLogger(logger *zap.Logger) ServiceOption {
    return func(s *service) {
        s.logger = logger
    }
}       
```

You can then use 0-n of these options when creating the service:

```
// default
NewService(logger, config, db)
```

```
// in tests
NewService(logger, config, db, WithHttpClient(client))
```